[CodeCompletion] Allow preChecking an expression twice in code completion

ahoppen opened the pull request:
In code completion we might call `preCheckExpression` twice - once for the first pass and once for the second pass. This is fine since `preCheckExpression` idempotent, so don't assert if we are in code completion mode.

Fixes rdar://79136653 [SR-14755]

ahoppen commented:
@swift-ci Please smoke test

rintaro commented:
@xedin Should `preCheckExpression()` be idempotent (https://github.com/apple/swift/pull/33749#discussion_r483774638)?

xedin commented:
I think it should be yes, although there might be edge cases which cases these crashes...

xedin commented:
Is this some kind of a stale pointer problem where sequence has been folded but something still holds a reference to an original version?

rintaro commented:
@xedin Yeah, [some expression kinds e.g. `AssignExpr` in SequenceExpr are mutated](https://github.com/apple/swift/blob/ea146ac5d4237e3f517b91584c018893158ae579/lib/Sema/TypeCheckExpr.cpp#L356-L386) in precheckExpression()

rintaro commented:
Maybe we can just disable these assertions (`!ifExpr->isFolded()`) when we are in code completion mode?

xedin commented:
I think it would be okay. Real solution here would be to type-check multi-statement closures together with their context just like single-expression ones, once that's done we should be able to remove all this complicated fallback logic.

ahoppen committed:
[CodeCompletion] Allow preChecking an expression twice in code completion

In code completion we might call `preCheckExpression` twice - once for the first pass and once for the second pass. This is fine since `preCheckExpression` idempotent, so don't assert if we are in code completion mode.

Fixes rdar://79136653 [SR-14755]

ahoppen commented:
OK, that’s the even simpler fix. I disabled the asserts if `CompletionCallback` is set.

ahoppen commented:
@swift-ci Please smoke test
